src/helpers/asn1.php

```diff
@@ -0,0 +1,259 @@
+                          'oct'=>'04',
+                          'obj'=>'06',
+                          'bit'=>'03',
+                          'printable'=>'13',
+                          'int'=>'02',
+                          'set'=>'31',
+                          'expl'=>'a',
+                          'utime'=>'17',
+                          'gtime'=>'18',
+                          'utf8'=>'0c',
+                          'ia5'=>'16',
+                          'visible'=>'1a',
+                          't61'=>'14',
+                          'impl'=>'80',
+                          'other'=>''
+                          );
+    if(array_key_exists($name, $functionList)) {
+      return $functionList[$name];
+    } else {
+      // echo "func \"$name\" not available";
+      return false;
+    }
+  }
+  
+  public static function __callStatic($func, $params) {
+    $func = strtolower($func);
+      $asn1Tag = self::asn1Tag($func);
+      if($asn1Tag !== false){
+        $num = $asn1Tag; //valu of array
+        $hex = $params[0];
+        $val = $hex;
+        if($func == 'printable' || $func == 'utf8' || $func == 'ia5' || $func == 'visible' || $func == 't61') { // ($string)
+          $val = bin2hex($hex);
+        }
+        if($func == 'int') {
+          $val = (strlen($val)%2 != 0)?"0$val":"$val";
+        }
+        if($func == 'expl') { //expl($num, $hex)
+          $num = $num.$params[0];
+          $val = $params[1];
+        }
+        if($func == 'impl') { //impl($num="0")
+          $val = (!$val)?"00":$val;  
+          $val = (strlen($val)%2 != 0)?"0$val":$val;
+          return $num.$val;
+        }
+        if($func == 'other') { //OTHER($id, $hex, $chr = false)
+          $id = $params[0];
+          $hex = $params[1];
+          $chr = @$params[2];
+          $str = $hex;
+          if($chr != false) {
+            $str = bin2hex($hex);
+          }
+          $ret = "$id".self::asn1_header($str).$str;
+          return $ret;
+        }
+        if($func == 'utime') {
+          $time = $params[0]; //yymmddhhiiss
+          $oldTz = date_default_timezone_get();
+          date_default_timezone_set("UTC");
+          $time = date("ymdHis", $time);
+          date_default_timezone_set($oldTz);
+          $val = bin2hex($time."Z");
+        }
+        if($func == 'gtime') {
+          if(!$time = strtotime($params[0])) {
+             // echo "asn1::GTIME function strtotime cant recognize time!! please check at input=\"{$params[0]}\"";
+            return false;
+          }
+          $oldTz = date_default_timezone_get();
+          // date_default_timezone_set("UTC");
+          $time = date("YmdHis", $time);
+          date_default_timezone_set($oldTz);
+          $val = bin2hex($time."Z");
+        }
+        $hdr = self::asn1_header($val);
+        return $num.$hdr.$val;
+    } else {
+      // echo "asn1 \"$func\" not exists!";
+    }
+  }
+  // =====End ASN.1 Builder section=====
+}
+?>
```

Comment:
Just a suggestion, in PHP you don't need to close when it's just a class file, you only need to leave a line break as end of file
This does not affect the code in any way, it is just a standard (it seems to me)

```diff
-?>
+
```

Reply:
Thanks. And does it also apply to bracket for `if` `else` in eg: pdfsign.php?
sometime its makes me confusing to find the boundary.

Reply:
On `if` `else` you can ignore the square brackets when they are single-line sentences. 
but the standard is to always put them